### PR TITLE
[jdbc]Add a new Kyuubi jdbc connector

### PR DESCRIPTION
**Pull request by chengduhaha**
## What changes were proposed in this pull request?

Create a new Kyuubi JDBC connector so that users may utilize Kyuubi to query Spark SQL and fill up Kyuubi JDBC in the hue.ini configuration file.
such as :
[[[kyuubi_jdbc]]]
   name=Kyuubi_JDBC
   interface=jdbc
   options='{"url": "jdbc:hive2://xxxx:10009", "driver": "org.apache.kyuubi.jdbc.KyuubiHiveDriver", "user": "xxxx", "password": "xxxx"}'
   

## How was this patch tested?
unit tests

Please review [Hue Contributing Guide](https://github.com/cloudera/hue/blob/master/CONTRIBUTING.md) before opening a pull request.

### CODE REVIEW

**Comment by Harshg999** [comment]
Hey @chengduhaha, thanks for your PR for Kyuubi support using JDBC interface! We'll review it shortly.

Can you also please share the steps to create a test setup (kyuubi + sparksql) and the test scenarios you performed? That'll be very helpful!

**Comment by chengduhaha** [comment]
> Hey @chengduhaha, thanks for your PR for Kyuubi support using JDBC interface! We'll review it shortly.
> 
> Can you also please share the steps to create a test setup (kyuubi + sparksql) and the test scenarios you performed? That'll be very helpful!

Sure, here is my  test step:


1. Use the instructions at https://kyuubi.readthedocs.io/en/master/quick_start/ to install and launch Kyuubi.
2. Go to https://kyuubi.readthedocs.io/en/master/client/jdbc/kyuubi_jdbc.html to get the Kyuubi JDBC driver.
3. Update hue.ini with the Kyuubi JDBC interface.
4. Restart Hue after adding the Kyuubi JDBC driver path to your Java CLASSPATH.
5. Use the Hue Kyuubi JDBC Editor to run sparksql queries.

Testing Scenarios:
1.Test running various SQL queries from Hue，this includes basic DDL DML queries.
2.Test with multiple users or queries concurrently to ensure stability and proper resource management.

**Comment by chengduhaha** [comment]
> Hey @chengduhaha, thanks for your PR for Kyuubi support using JDBC interface! We'll review it shortly.
> 
> Can you also please share the steps to create a test setup (kyuubi + sparksql) and the test scenarios you performed? That'll be very helpful!

Hi @Harshg999 ,
Thank you for reviewing the pull request. I appreciate your interest in the setup and testing details. Here's a brief overview of how to set up a test environment for Kyuubi with Spark SQL and the testing scenarios I performed:

Setting up Kyuubi + Spark SQL Test Environment:
1.Use the instructions at https://kyuubi.readthedocs.io/en/master/quick_start/ to install and launch Kyuubi.
2.Go to https://kyuubi.readthedocs.io/en/master/client/jdbc/kyuubi_jdbc.html to get the Kyuubi JDBC driver.
3.Update hue.ini with the Kyuubi JDBC interface.
4.Restart Hue after adding the Kyuubi JDBC driver path to your Java CLASSPATH.
5.Use the Hue Kyuubi JDBC Editor to run sparksql queries.

Testing Scenarios:
1.Test running various SQL queries from Hue，this includes basic DDL DML queries.
2.Test with multiple users or queries concurrently to ensure stability and proper resource management.

**Comment by chengduhaha** [comment]
Dear @bjornalm, @agl29, and @Harshg999,

I have submitted a pull request (PR) and require at least two approving reviews for this request to be merged. Please review my code at your earliest convenience and provide feedback or approval. Thank you very much for your time and assistance.

Sincerely,
chengduhaha

File: desktop/libs/notebook/src/notebook/connectors/jdbc_kyuubi.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from librdbms.jdbc import query_and_fetch
+
+from notebook.connectors.jdbc import JdbcApi
+from notebook.connectors.jdbc import Assist
+class JdbcApiKyuubi(JdbcApi):
```
Comment (Harshg999) [inline desktop/libs/notebook/src/notebook/connectors/jdbc_kyuubi.py:22]: Nit: line space

**Review by Harshg999 — APPROVED** [review]
Hi @chengduhaha, last week was holiday off so I couldn't test this PR. 
Today, I was successful in spinning a test setup and verify the Hue-Kyuubi via JDBC interface.

The change looks good to me, thank you for your contribution!

**Comment by Harshg999** [comment]
@chengduhaha Would you like to take up adding the Hue-Kyuubi JDBC config in the Hue documentation too? :)

You can try checking how JDBC config doc section is added for other connectors in https://docs.gethue.com/administrator/configuration/connectors/ and open a separate PR for a docs change.

File: desktop/libs/notebook/src/notebook/connectors/jdbc_kyuubi.py [hunk 2]
```diff
@@ -19,6 +19,7 @@
 
 from notebook.connectors.jdbc import JdbcApi
 from notebook.connectors.jdbc import Assist
+
```
Comment (chengduhaha) [inline desktop/libs/notebook/src/notebook/connectors/jdbc_kyuubi.py:22]: @Harshg999 I added line space after import, thanks.

File: desktop/libs/notebook/src/notebook/connectors/jdbc_kyuubi.py [hunk 3]
```diff
@@ -19,6 +19,7 @@
...
 class JdbcApiKyuubi(JdbcApi):
```
Comment (chengduhaha) [inline desktop/libs/notebook/src/notebook/connectors/jdbc_kyuubi.py:23]: @Harshg999 I added line space after import, thanks.

**Comment by chengduhaha** [comment]
Hi @Harshg999,

I wanted to extend my gratitude for your insightful review and valuable suggestions on my previous PR. I have made the recommended modifications and pushed the updated code.

Additionally, following your suggestion, I've created a new PR to add the Kyuubi JDBC config to the Hue documentation. The PR can be found here: [Add Kyuubi JDBC Config to Hue Documentation #3593](https://github.com/cloudera/hue/pull/3593).

I'm looking forward to any feedback you might have on this new contribution.

Thanks again for your guidance and support!

**Review by agl29 — APPROVED** [review]
Looks good to me!
Thanks @chengduhaha for the contribution.

**Comment by 2416210017** [comment]
hi, @chengduhaha ,I have already installed the previous steps, may I ask why there is an error when using hue to query kyuubi?

```
# kyuubi
 echo 'export KYUUBIL_HOME=/usr/share/hue/desktop/kyuubi' >> /etc/profile
 echo 'export CLASSPATH=$KYUUBIL_HOME/kyuubi-hive-jdbc-shaded-1.7.3.jar:$CLASSPATH' >> /etc/profile
````
restart hue_latest_test

image

Query error:

image

**Comment by 2416210017** [comment]
Here's a question,Hue not validating user credentials and just caching . Need to restart hue service to again prompt for login prompt if wrong credentials entered.（https://github.com/cloudera/hue/issues/3529）

**Pull request merged**